In this exercise, you will practice basic `dplyr` (the relevant package within `tidyverse`) operations on a dataset of gene expression levels across human tissues (from the GTEx Project). The dataset (`dt`) has one row per gene, with columns:

-   `Name`: the Ensembl gene ID\
-   `Description`: the gene symbol
-   One column per tissue containing the median expression levels (in TPM units)

------------------------------------------------------------------------

### Start by loading the data into R:

Note the arguments, which specify that the first two lines of the file will be skipped, that the file contains a header line that should be used to form column names, and that the columns are tab (`\t`) separated.

```{r}
library(tidyverse)
dt <- read_delim("GTEx_Analysis_2017-06-05_v8_RNASeQCv1.1.9_gene_median_tpm.gct", skip = 2, col_names = TRUE, delim = "\t")
```

### Q1. How many genes are in the dataset, and how many tissues are represented?

```{r}

# code and answer here
gene_count <- nrow(dt)
# 56200 genes

tissue_count <- ncol(dt)-2
# 54

```

------------------------------------------------------------------------

### Q2. How many genes are expressed above a given threshold in each tissue?

A common threshold is 1 TPM to define whether a gene is "expressed."\
Compute, for each of the three tissues, `Liver`, `Testis`, and `Whole Blood`, how many genes are expressed above this threshold.\
Then identify which tissue has the **most** and which has the **fewest** expressed genes. See this paper for one possible explanation: Xia et al. 2021: [PMC7891839](https://pmc.ncbi.nlm.nih.gov/articles/PMC7891839/).

Do the results change if you set the threshold higher, say at 10 TPM?

```{r}

# Code here
# Thresh = 1 TPM
liver <- select(dt,Liver)
liver_thresh1 = length(liver[liver > 1])
# 13499

testis <- select(dt,Testis)
testis_thresh1 <- length(testis[testis > 1])
# 25064

wb <- select(dt,"Whole Blood")
wb_thresh1 <- length(wb[wb > 1])
# 11616

# Thresh = 10 TPM
liver_thresh10 <- length(liver[liver > 10])
# 5132

testis_thresh10 <- length(testis[testis > 10])
# 11471

wb_thresh10 <- length(wb[wb > 10])
# 4482

# Interpretation here as a comment
# Testis has the most expressed genes for both thresholds (1 and 10 TPM), while WB has the fewest. Increasing the TPM threshold decreases the counts of expressed genes across all three tissues as expected. However, the ranking of the tissues from fewest to most expressed genes does not change.

```

------------------------------------------------------------------------

### Q3. What are the 20 most highly expressed genes in the liver?

What do you notice about many of the genes in this list? Can you propose a hypothesis to explain your observation?

```{r}

# Code here
liver_sorted <- dt %>% arrange(desc(Liver)) %>% select(Name,Description,Liver) %>% slice(1:20)

# Interpretation here as a comment
# 60% of the genes are mitochondrial genes. High expression of mitochondrial genes is due to the energy demands of the primary functions of the liver, including glucose, lipid, and protein metabolism and detoxification. 

```

------------------------------------------------------------------------

### Q4. Compare the mean expression levels of mitochondrial versus nuclear-encoded genes in the Liver.

Genes encoded by the mitochondrial genome have the gene symbol prefix `MT-`. The function `str_starts()` from the `stringr` package in `tidyverse` tests whether a string (or a vector of strings) start with a given pattern, returning TRUE/FALSE (or a boolean vector) as output.

Use `str_starts()` to identify which genes are mitochondrially encoded and store this information in a new Boolean column named `is_mitochondrial`. Next, compute the mean and median expression of mitochondrial genes in the liver, as well as the mean and median expression of nuclear-encoded genes in the liver and compare them. Do mitochondrial genes exhibit higher or lower mean and median expression than nuclear genes? What does the difference between the mean and median tell you about the distribution of gene expression.

```{r}
# Code here

dt_mito <- dt %>% mutate(is_mitochondrial = str_starts(Description,"MT-")) # Brooke W. forced me to write it this way at gunpoint

mito_med <- dt_mito %>% group_by(is_mitochondrial) %>% summarize(mean = mean(Liver), median(Liver))

# Interpretation here as a comment
# Nuclear: mean = 8.37, median = 0
# Mito: mean = 11592.72, median = 3.4
# The nuclear expression is much lower than the mitochondrial expression, and the distribution is relatively even skewed, indicating that most nuclear genes have a low level of expression around the mean. The mitochondrial expression is right skewed, indicating that the majority of mitochondrial genes have high expression with a few outliers with lower expression.

```

------------------------------------------------------------------------

### Q5. Are liverâ€™s most highly expressed genes also highly expressed in another tissue?

Take the 20 most highly expressed genes in **Liver** and check how they are expressed in the **Heart - Left Ventricle**. Identify:\
1. one gene that is highly expressed in both tissues\
2. one gene that is highly expressed in Liver but very low in Heart

Look up basic information about these genes on the internet. Do their expression patterns make sense given their known functions?

```{r}

# Code here
liver_HLV_sorted <- dt %>% arrange(desc(Liver)) %>% select(Name,Description,Liver,"Heart - Left Ventricle") %>% slice(1:20)

# 1.
liver_HLV_sorted <- liver_HLV_sorted %>% rename(HLV = "Heart - Left Ventricle")
both_high <- liver_HLV_sorted %>% arrange(desc(HLV)) %>% slice(1)
# MT-CO1

# 2.
one_high <- liver_HLV_sorted %>% arrange(desc(HLV)) %>% slice(20)
# FBG

# Interpretation here as a comment
# MT-CO1 transfers electrons from cytochrome c to oxygen and is an important part of the electron transport chain, which is widely used in all tissues.
# FBG is produced in the liver, which is why it has a much higher expression than in the heart.

```

------------------------------------------------------------------------

### Q6. Comparing gene expression between tissues

Another way to compare tissues is to ask how similar are their expression patterns overall, across all genes?

R has a built-in function called `cor.test()` that measures the correlation between two vectors of numbers. Remember that you can extract the contents of a given column of a tibble as a vector using the syntax `tibble_name$column_name`. If a column name contains spaces, enclose the column name in backticks such as `` tibble_name$`column name` ``.

Use `cor.test()` with the argument `method = "kendall"` to compute: - the Spearman correlation between `Liver` and `Heart - Left Ventricle` - the Spearman correlation between `Liver` and `Brian - Cortex`

Which pair of tissues is more strongly correlated? Does this result make sense biologically?

```{r}
# Liver vs Heart answer
LHB <- dt %>% select(Liver,"Heart - Left Ventricle","Brain - Cortex") %>% rename(HLV = "Heart - Left Ventricle") %>% rename(BC = "Brain - Cortex")

L = LHB$Liver
H = LHB$HLV
B = LHB$BC

LH_cor = cor.test(L,H,method = "kendall")

# Liver vs Brain answer

LB_cor = cor.test(L,B,method = "kendall")

# interpretation
# Liver and heart has a stronger Spearman correlation than liver and brain. The brain performs vastly different functions than the liver and heart, which could be one reason that there is less correlation in gene expression.

```

------------------------------------------------------------------------

### Optional: further exploration

What other patterns are present in the data? Use tidyverse functions to explore and summarize the data in different ways and see what you can learn.

```{r}
# answer here
```
